### PR TITLE
generate crash reports

### PR DESCRIPTION
**Pull request by mwhudson**
This doesn't have any of the nice UI we designed for presenting and uploading crash reports, but it should still be useful. Please ask if anything doesn't make sense, some of the mysterious things might become clearer when I add the UI code but quite possibly some of it is redundant now.

### CODE REVIEW

File: subiquity/controllers/error.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+class ErrorReportState(enum.Enum):
+    INCOMPLETE = _("INCOMPLETE")
+    DONE = _("DONE")
+    ERROR = _("ERROR")
+
+
+class ErrorReportKind(enum.Enum):
+    BP_FAIL = _("Block device probe failure")
+    RBP_FAIL = _("Disk probe failure")
```
Comment (raharper) [inline subiquity/controllers/error.py]: BLOCK_PROBE_FAIL
DISK_PROBE_FAIL

Would make reading the enums in the rest of the code easier.

File: subiquity/controllers/error.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    meta = attr.ib(default=attr.Factory(dict))
+
+    @classmethod
+    def new(cls, controller, kind):
+        prefix = "installer.{}.{}".format(
+            datetime.datetime.utcnow().isoformat(timespec='seconds'),
+            kind.name.lower())
+        i = 0
+        while 1:
```
Comment (raharper) [inline subiquity/controllers/error.py]: This should have some sort of upper-bound and timeout.

File: subiquity/controllers/filesystem.py [hunk 3]
```diff
@@ -64,6 +65,11 @@ def __init__(self, controller, restricted, timeout, cb):
         self.cb = cb
         self.state = ProbeState.NOT_STARTED
         self.result = None
+        if restricted:
+            self.kind = ErrorReportKind.RBP_FAIL
+        else:
+            self.kind = ErrorReportKind.BP_FAIL
```
Comment (raharper) [inline subiquity/controllers/filesystem.py]: I see two kinds here, but below there are 3 scenarios (bp failed, bp timeout, bp data loading fail)  Should each be a separate enum?

File: subiquity/controllers/error.py (hunk 1 above)
Comment (mwhudson) [reply subiquity/controllers/error.py]: OK

File: subiquity/controllers/error.py (hunk 2 above)
Comment (mwhudson) [reply subiquity/controllers/error.py]: How about an upper bound of 10k? :)

File: subiquity/controllers/filesystem.py (hunk 3 above)
Comment (mwhudson) [reply subiquity/controllers/filesystem.py]: Hmm I tend to think not. It only really makes sense to have a different kind if we want to display them differently in the UI and I don't think that we really will. Can always change this later when I add the UI if my thinking changes here.

File: subiquity/controllers/error.py [hunk 4]
```diff
@@ -61,7 +61,7 @@ def new(cls, controller, kind):
             datetime.datetime.utcnow().isoformat(timespec='seconds'),
             kind.name.lower())
         i = 0
-        while 1:
+        while i < 10000:
```
Comment (raharper) [inline subiquity/controllers/error.py]: I guess I don't understand the failing scenarios.  Is subiquity waiting on something and it keeps trying until it's present?  If we're waiting, shouldn't we sleep in this loop?  If we're not sleeping in this loop will we hit 10k too fast?

Comment (mwhudson) [reply subiquity/controllers/error.py]: No, this is just the usual thing of atomically opening a file without overwriting an existing one. You do know what the 'x' flag means to open()?

Comment (raharper) [reply subiquity/controllers/error.py]: OK, but I don't understand the failure scenario.  Multiple crashes in the same subiquity run?  Surely using a time-based filename should be sufficient, something with micro-second resolution won't end up overlapping multiple calls.

Comment (mwhudson) [reply subiquity/controllers/error.py]: To some extent this is perhaps overthought as the expected number of crash files produced in any given run is at most 2 (of different kinds, to boot). But I do want the filenames to be human-friendly: they are saved to the install media by default, and if the user then pops their usb stick in their desktop to report a bug from there, I want them to be able to find them.

I guess I don't understand your concern here. The fact that this code is complex? That it has undesired consequences?

Comment (mwhudson) [reply subiquity/controllers/error.py]: BTW the most recent version just puts the date in the timestamp.

**Comment by mwhudson** [comment]
I've redone this, putting a nano-second level timestamp into the filename instead of looping. I've rebased rather than made new commits which might make re-review a bit trickier. Sorry!

**Pull request merged**